File: LeetCode/Linked List/Easy/1290. Convert Binary Number in a Linked List to Integer/main.cpp
#include <iostream>
using namespace std;

// Definition for singly-linked list.// just for testing
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    int getDecimalValue(ListNode *head)
    {
        int ans = 0;
        while (head)
        {
            ans = ans * 2 + head->val;
            head = head->next;
        }
        return ans;
    }
};

// just for testing
int main()
{
    Solution s;
    // for a linked list with 101 binary values in diff nodes
    ListNode *head = new ListNode(0, new ListNode(0, new ListNode(1)));
    int ans = s.getDecimalValue(head);
    cout << ans;
}

File: LeetCode/Linked List/Easy/1290. Convert Binary Number in a Linked List to Integer/main.py
# Definition for singly-linked list. # for testing
class ListNode():
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


"""
we conevert binary to decimal by multiplying existing sum by 2 and adding the liked list binary value
as shown in the picture
"""


class Solution():
    def getDecimalValue(self, head):
        sum = 0
        while head:
            sum = 2*sum + head.val
            head = head.next
        return sum


# For testing
x = ListNode(1, ListNode(0, ListNode(1)))
obj = Solution()
print(obj.getDecimalValue(x))
